rootfs/apply.go

```diff
@@ -47,76 +48,123 @@ type Layer struct {
-			return "", err
+	chain := make([]digest.Digest, len(layers))
+	for i, layer := range layers {
+		chain[i] = layer.Diff.Digest
+	}
+	chainID := identity.ChainID(chain)
+
+	var tries int
+	for {
+		// Just stat top layer, remaining layers will have their existence checked
+		// on prepare. Calling prepare on upper layers first guarantees that upper
+		// layers are not removed while calling stat on lower layers
+		_, err := sn.Stat(ctx, chainID.String())
+		if err != nil {
+			tries++
+			if tries < 3 && errdefs.IsNotFound(err) {
+				if err := applyLayers(ctx, layers, chain, sn, a); err != nil {
+					if errdefs.IsUnavailable(err) {
+						time.Sleep(time.Second * time.Duration(tries))
+					}
+					if !errdefs.IsAlreadyExists(err) {
+						return "", err
+					}
+					continue
+				}
+			} else {
+				return "", errors.Wrapf(err, "failed to stat snapshot %s", chainID)
+			}
 		}
-
-		chain = append(chain, layer.Diff.Digest)
+		break
 	}
-	return identity.ChainID(chain), nil
+
+	return chainID, nil
 }
 
 // ApplyLayer applies a single layer on top of the given provided layer chain,
 // using the provided snapshotter and applier. If the layer was unpacked true
 // is returned, if the layer already exists false is returned.
 func ApplyLayer(ctx context.Context, layer Layer, chain []digest.Digest, sn snapshots.Snapshotter, a diff.Applier, opts ...snapshots.Opt) (bool, error) {
+	var applied bool
+	err := applyLayers(ctx, []Layer{layer}, append(chain, layer.Diff.Digest), sn, a, opts...)
+	if err != nil {
+		if !errdefs.IsAlreadyExists(err) {
+			return false, err
+		}
+	} else {
+		applied = true
+	}
+	return applied, nil
+}
+
+func applyLayers(ctx context.Context, layers []Layer, chain []digest.Digest, sn snapshots.Snapshotter, a diff.Applier, opts ...snapshots.Opt) error {
 	var (
-		parent  = identity.ChainID(chain)
-		chainID = identity.ChainID(append(chain, layer.Diff.Digest))
+		parent  = identity.ChainID(chain[:len(chain)-1])
+		chainID = identity.ChainID(chain)
+		layer   = layers[len(layers)-1]
 		diff    ocispec.Descriptor
+		mounts  []mount.Mount
+		err     error
 	)
 
-	_, err := sn.Stat(ctx, chainID.String())
-	if err == nil {
-		log.G(ctx).Debugf("Extraction not needed, layer snapshot %s exists", chainID)
-		return false, nil
-	} else if !errdefs.IsNotFound(err) {
-		return false, errors.Wrapf(err, "failed to stat snapshot %s", chainID)
-	}
-
 	key := fmt.Sprintf("extract-%s %s", uniquePart(), chainID)
 
-	// Prepare snapshot with from parent, label as root
-	mounts, err := sn.Prepare(ctx, key, parent.String(), opts...)
-	if err != nil {
-		//TODO: If is snapshot exists error, retry
-		return false, errors.Wrapf(err, "failed to prepare extraction snapshot %q", key)
+	var tries int
+	for {
+		// Prepare snapshot with from parent, label as root
+		mounts, err = sn.Prepare(ctx, key, parent.String(), opts...)
+		if err != nil {
+			tries++
+			if tries < 3 && len(layers) > 1 && errdefs.IsNotFound(err) {
```

Comment:
this is kinda complex, do we need the `len(layers)` part in this?

Reply:
That `len(layers)` check is to prevent recursion when called from `rootfs.ApplyLayer`. I tried to unify the functions, let me see if I can do this in a less complex well, else will document that.